### PR TITLE
feat: reduce cognitive complexity

### PR DESCRIPTION
**Pull request by jean-michelet**
This PR is intended to reduce the overall cognitive complexity of the component, which is imho unnecessarily hard to read, although I understand this is subjective. Basically, I've done a bit of refactoring, breaking down huge functions into smaller ones and limiting the use of nested closures.

I put the `findPlugins` function in its own file to reduce the size of `index.js` and because it is a clearly separated step.

I tried to do the same with `loadPlugins` but `vitest` failed to transpile after that... But the two big steps are finding and loading, registering is just a few lines so I think it's ok.

I think I've also slightly reduced algorithmic complexity by making a single global iteration on `pluginTree`, including plugin loading, hooks and registration. This also makes it possible to benefit from the current iteration context, instead of making distinct tree traversal for hooks and plugins, although this point isn't quite clear to me yet. So it's not a fix of #383.

### CODE REVIEW

File: find-plugins.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
+const { readdir } = require('node:fs/promises')
```
Comment (mcollina) [inline find-plugins.js:1]:
```suggestion
'use strict'

const { readdir } = require('node:fs/promises')
```

File: find-plugins.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+const { relative, join } = require('path')
+const runtime = require('./runtime')
```
Comment (mcollina) [inline find-plugins.js:3]: can you move this file and `runtime` into `lib/`?

File: find-plugins.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    if (dirent.isDirectory() && !atMaxDepth) {
+      processDirectory({ prefix, opts, dirent, dir, file, directoryPromises, hookedAccumulator, depth, currentHooks })
+    } else if (indexDirent) {
+      // An index.js file is present in the directory so we ignore the others modules (but not the subdirectories)
+    } else if (dirent.isFile() && opts.scriptPattern.test(dirent.name)) {
+      processFile({ file, opts, dirent, hookedAccumulator, prefix })
+    }
+  }
+
+  await Promise.all(directoryPromises)
```
Comment (mcollina) [inline find-plugins.js:95]: Can you avoid having a `directoryPromises` array at all? I think this is one of the things that makes this code very hard to read. We should be making `processDirectory` `async` instead and `await` it in place.

Comment (jean-michelet) [reply find-plugins.js]: Yes.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by jean-michelet** [comment]
Just look for my last commit plz, I had missed a previous optimization during refactorisation.

**Review by gurgunday — APPROVED** [review]
LGTM

**Comment by gurgunday** [comment]
Can you update the workflow to version 5?

https://github.com/fastify/fastify-autoload/blob/3584972aa271aed85c4443b759f8c53a640db773/.github/workflows/ci.yml#L20

**Comment by jean-michelet** [comment]
I do it in an independent PR @gurgunday : https://github.com/fastify/fastify-autoload/pull/403

**Comment by jean-michelet** [comment]
I would prefer a review from @climba03003 before merging.

**Comment by jean-michelet** [comment]
Can you review @climba03003 plz or can we merge?

**Review by climba03003 — APPROVED** [review]
LGTM.

**Pull request merged**